File: src/sttp/sttp-core/Application/Services/SyncCoordinatorService.cs
using Microsoft.Extensions.Logging;
using SttpMcp.Domain.Contracts;
using SttpMcp.Domain.Models;

namespace SttpMcp.Application.Services;

public sealed class SyncCoordinatorService(
    INodeStore store,
    ISyncChangeSource source,
    ILogger<SyncCoordinatorService> logger,
    ISyncCoordinatorPolicy? policy = null)
{
    public async Task<SyncPullResult> PullAsync(
        SyncPullRequest request,
        CancellationToken ct = default)
    {
        var pageSize = Math.Clamp(request.PageSize, 1, 500);
        var maxBatches = Math.Max(request.MaxBatches ?? int.MaxValue, 1);
        var checkpoint = await store.GetCheckpointAsync(request.SessionId, request.ConnectorId, ct);
        var cursor = checkpoint?.Cursor;
        var fetched = 0;
        var created = 0;
        var updated = 0;
        var duplicate = 0;
        var skipped = 0;
        var filtered = 0;
        var batches = 0;
        var hasMore = false;

        while (batches < maxBatches)
        {
            var page = await source.ReadChangesAsync(
                request.SessionId,
                request.ConnectorId,
                cursor,
                pageSize,
                ct);

            if (page.Nodes.Count == 0)
            {
                hasMore = page.HasMore;
                break;
            }

            batches++;
            fetched += page.Nodes.Count;
            SttpNode? lastAppliedNode = null;

            foreach (var node in page.Nodes)
            {
                if (!(policy?.ShouldAcceptNode(node) ?? true))
                {
                    filtered++;
                    continue;
                }

                var upsert = await store.UpsertNodeAsync(node, ct);
                lastAppliedNode = node;

                switch (upsert.Status)
                {
                    case NodeUpsertStatus.Created:
                        created++;
                        break;
                    case NodeUpsertStatus.Updated:
                        updated++;
                        break;
                    case NodeUpsertStatus.Duplicate:
                        duplicate++;
                        break;
                    case NodeUpsertStatus.Skipped:
                        skipped++;
                        break;
                }
            }

            if (page.NextCursor is not null)
            {
                cursor = page.NextCursor;
                checkpoint = new SyncCheckpoint
                {
                    SessionId = request.SessionId,
                    ConnectorId = request.ConnectorId,
                    Cursor = page.NextCursor,
                    UpdatedAt = DateTime.UtcNow,
                    Metadata = policy?.BuildCheckpointMetadata(
                            request.SessionId,
                            request.ConnectorId,
                            checkpoint,
                            lastAppliedNode,
                            page.NextCursor)
                        ?? checkpoint?.Metadata
                };

                await store.PutCheckpointAsync(checkpoint, ct);
            }

            hasMore = page.HasMore;
            if (!hasMore)
                break;
        }

        logger.LogDebug(
            "Sync pull completed for {SessionId}/{ConnectorId}: fetched={Fetched}, created={Created}, updated={Updated}, duplicate={Duplicate}, skipped={Skipped}, filtered={Filtered}, batches={Batches}, hasMore={HasMore}",
            request.SessionId,
            request.ConnectorId,
            fetched,
            created,
            updated,
            duplicate,
            skipped,
            filtered,
            batches,
            hasMore);

        return new SyncPullResult
        {
            Fetched = fetched,
            Created = created,
            Updated = updated,
            Duplicate = duplicate,
            Skipped = skipped,
            Filtered = filtered,
            Batches = batches,
            HasMore = hasMore,
            LastCursor = cursor,
            Checkpoint = checkpoint
        };
    }
}

File: src/sttp/sttp-core-rs/src/application/services/sync_coordinator_service.rs
use std::sync::Arc;

use anyhow::Result;
use chrono::Utc;

use crate::domain::contracts::{NodeStore, SyncChangeSource, SyncCoordinatorPolicy};
use crate::domain::models::{
    ConnectorMetadata, NodeUpsertStatus, SttpNode, SyncCheckpoint, SyncCursor,
    SyncPullRequest, SyncPullResult,
};

pub struct SyncCoordinatorService {
    store: Arc<dyn NodeStore>,
    source: Arc<dyn SyncChangeSource>,
    policy: Option<Arc<dyn SyncCoordinatorPolicy>>,
}

impl SyncCoordinatorService {
    pub fn new(store: Arc<dyn NodeStore>, source: Arc<dyn SyncChangeSource>) -> Self {
        Self {
            store,
            source,
            policy: None,
        }
    }

    pub fn with_policy(
        store: Arc<dyn NodeStore>,
        source: Arc<dyn SyncChangeSource>,
        policy: Arc<dyn SyncCoordinatorPolicy>,
    ) -> Self {
        Self {
            store,
            source,
            policy: Some(policy),
        }
    }

    pub async fn pull_async(&self, request: SyncPullRequest) -> Result<SyncPullResult> {
        let page_size = request.page_size.clamp(1, 500);
        let max_batches = request.max_batches.unwrap_or(usize::MAX).max(1);
        let mut checkpoint = self
            .store
            .get_checkpoint_async(&request.session_id, &request.connector_id)
            .await?;
        let mut cursor = checkpoint.as_ref().and_then(|value| value.cursor.clone());
        let mut result = SyncPullResult::default();

        while result.batches < max_batches {
            let page = self
                .source
                .read_changes_async(
                    &request.session_id,
                    &request.connector_id,
                    cursor.clone(),
                    page_size,
                )
                .await?;

            if page.nodes.is_empty() {
                result.has_more = page.has_more;
                break;
            }

            result.batches += 1;
            result.fetched += page.nodes.len();

            let mut last_applied_node: Option<SttpNode> = None;

            for node in page.nodes {
                if !self.should_accept_node(&node) {
                    result.filtered += 1;
                    continue;
                }

                let upsert = self.store.upsert_node_async(node.clone()).await?;
                last_applied_node = Some(node);

                match upsert.status {
                    NodeUpsertStatus::Created => result.created += 1,
                    NodeUpsertStatus::Updated => result.updated += 1,
                    NodeUpsertStatus::Duplicate => result.duplicate += 1,
                    NodeUpsertStatus::Skipped => result.skipped += 1,
                }
            }

            if let Some(next_cursor) = page.next_cursor {
                cursor = Some(next_cursor.clone());
                let next_checkpoint = SyncCheckpoint {
                    session_id: request.session_id.clone(),
                    connector_id: request.connector_id.clone(),
                    cursor: Some(next_cursor.clone()),
                    updated_at: Utc::now(),
                    metadata: self.build_checkpoint_metadata(
                        &request.session_id,
                        &request.connector_id,
                        checkpoint.as_ref(),
                        last_applied_node.as_ref(),
                        Some(&next_cursor),
                    ),
                };

                self.store.put_checkpoint_async(next_checkpoint.clone()).await?;
                checkpoint = Some(next_checkpoint);
            }

            result.has_more = page.has_more;
            if !result.has_more {
                break;
            }
        }

        result.last_cursor = cursor;
        result.checkpoint = checkpoint;
        Ok(result)
    }

    fn should_accept_node(&self, node: &SttpNode) -> bool {
        self.policy
            .as_ref()
            .map(|policy| policy.should_accept_node(node))
            .unwrap_or(true)
    }

    fn build_checkpoint_metadata(
        &self,
        session_id: &str,
        connector_id: &str,
        previous: Option<&SyncCheckpoint>,
        last_applied_node: Option<&SttpNode>,
        next_cursor: Option<&SyncCursor>,
    ) -> Option<ConnectorMetadata> {
        self.policy
            .as_ref()
            .and_then(|policy| {
                policy.checkpoint_metadata(
                    session_id,
                    connector_id,
                    previous,
                    last_applied_node,
                    next_cursor,
                )
            })
            .or_else(|| previous.and_then(|checkpoint| checkpoint.metadata.clone()))
    }
}